### PR TITLE
Pytorch transformer (take 2)

### PR DESCRIPTION
**Pull request by ShigekiKarita**
this is rework of #555 with upstream chainer implementation #655 on master

- [x] split modules as previous discussion in #555 #655
- [x] add docstrings
- [x] update asr train and recog with transformer
- [x] implement dynamic module loading like preprocessing transform module (also in chainer)
- [x] CTC/LM joint decoding
- [x] constent `ASRInterface` for all the E2E implementation
- [x] add pytorch exp and RESULTS (maybe finish tomorrow)

### CODE REVIEW

**Comment by ShigekiKarita** [comment]
## dynamic_import

Instead of @Fhrozen's `--model-module espnet.asr.pytorch_backend.e2e_asr_transformer` style, I prefer `--model-module espnet.asr.pytorch_backend.e2e_asr_transformer:E2E` because

- This notation is already used in `espnet.transform.transformation`
- It is less ambiguous. Users can easily understand what is the model.
- This `--model-module` does not only requires `E2E` class but also `PlotAttentionReport` class and `add_arguments`, `recognize` methods. Instead of those implicit requirements, I rather suggest `ASRInterface` for imported module. After `model_class = dynamic_import(args.model_module)`, it checks `assert isinstance(model, ASRInterface)`

## ASRInterface

This interface only has four methods that have been commonly used in several `E2E` class.
```python
class ASRInterface(object):
    @staticmethod
    def add_arguments(parser):
        return parser

    def recognize(self, x, recog_args, char_list=None, rnnlm=None):
        '''recognize x
        :param ndarray x: input acouctic feature (B, T, D) or (T, D)
        :param namespace recog_args: argment namespace contraining options
        :param list char_list: list of characters
        :param torch.nn.Module rnnlm: language model module
        :return: N-best decoding results
        :rtype: list
        '''
        pass

    def calculate_all_attentions(self, xs, ilens, ys):
        '''attention calculation
        :param list xs_pad: list of padded input sequences [(T1, idim), (T2, idim), ...]
        :param ndarray ilens: batch of lengths of input sequences (B)
        :param list ys: list of character id sequence tensor [(L1), (L2), (L3), ...]
        :return: attention weights (B, Lmax, Tmax)
        :rtype: float ndarray
        '''

    @property
    def attention_plot_class(self):
        from espnet.asr.asr_utils import PlotAttentionReport
            return PlotAttentionReport
```

There is still missing methods `__call__(self, xs, ilens, ys)` for chainer or `forward(self, xspad, ilens, yspad)` for pytorch but I would like leave this method for their differences at this time.

**Comment by ShigekiKarita** [comment]
I'm not familiar to chainer. Does the latest chainer.Chain (v5) prefer `forward` instead of `__call__`? https://docs.chainer.org/en/stable/guides/models.html
If it is true, I would add the `forward` method to `ASRInterface` and replace `__call__` with `forward`.

**Comment by Fhrozen** [comment]
Yep, I did not realized but there are accepting `forward` instead of `__call__`: https://github.com/chainer/chainer/blob/v5.0.0/examples/mnist/train_mnist.py

**Comment by ShigekiKarita** [comment]
Thanks. I will update it (maybe only E2E now).

**Comment by Fhrozen** [comment]
Ok, it accept `forward` and `__call__`: https://github.com/chainer/chainer/blob/e4a0b8fca2fdd893c491de79f0b66129d0d15824/chainer/link.py#L219

**Comment by ShigekiKarita** [comment]
I think my work is almost done. I'm just waiting experimental results. Maybe I will test multi GPUs later.

**Comment by ShigekiKarita** [comment]
https://github.com/espnet/espnet/blob/0afc759fd631d10c362ea3a2c537e7e54e1047a3/espnet/nets/asr_interface.py
I've added the `forward` method to `ASRInterface`.

**Comment by ShigekiKarita** [comment]
@Fhrozen Yeah, I also want to keep consistent between chainer and pytorch. I think this style is better. At the previous PR #655, I just forgot this comment
> I'd advise splitting as much as possible when it is not already merged, because if you decide to modularize later it will probably cause a lot of conflicts.    
The file is already 700+ lines long, and contains helper methods for the arguments and the transformer, classes for the encoder, decoder, attention, etc, so if they are not compatible with the RNN model, I'd say to put them in different files in `espnet.asr.pytorch_backend.transformer` for example.

_Originally posted by @gtache in https://github.com/espnet/espnet/pull/555_

**Comment by ShigekiKarita** [comment]
I've decoded intermediate Transformer+CTC (epoch 16/100) and RNNLM (epoch 5/20) for test. Result seems to be OK but I do not think this the best.

```
write a CER (or TER) result in exp/train_si284_pytorch_home/decode_test_dev93_beam10_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_home/result.txt                                                                                                     
|    SPKR      |   # Snt        # Wrd    |   Corr          Sub         Del          Ins         Err        S.Err    |        
|    Sum/Avg   |    503         48634    |   96.4          1.5         2.1          0.7         4.3         67.4    |        
write a WER result in exp/train_si284_pytorch_home/decode_test_dev93_beam10_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_home/result.wrd.txt                                                                                                          
|    SPKR      |    # Snt       # Wrd    |    Corr          Sub           Del          Ins          Err        S.Err    |    
|    Sum/Avg   |     503         8234    |    92.2          6.7           1.2          1.3          9.1         62.4    |

write a CER (or TER) result in exp/train_si284_pytorch_home/decode_test_eval92_beam10_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_home/result.txt                                                                                                    
|    SPKR      |   # Snt        # Wrd    |    Corr         Sub          Del          Ins         Err        S.Err    |       
|    Sum/Avg   |    333         33341    |    97.4         1.1          1.5          0.7         3.2         58.0    |       
write a WER result in exp/train_si284_pytorch_home/decode_test_eval92_beam10_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_home/result.wrd.txt                                                                                                         
|    SPKR      |    # Snt       # Wrd     |    Corr          Sub          Del          Ins           Err        S.Err    |   
|    Sum/Avg   |     333         5643     |    94.5          4.8          0.7          1.1           6.6         51.7    |
```

**Comment by ShigekiKarita** [comment]
The `patience=3` early stopped training. This is the result
```
write a CER (or TER) result in exp/train_si284_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu4_bs16_lr10.0_warmup25000_mli512_mlo150_epochs100_accum1_lennormFalse_lsmunigram0.1/decode_test_eval92_beam10_emodel.last10.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result.txt             
|               SPKR                  |               # Snt                               # Wrd                |               Corr                                 Sub                                Del                                 Ins                                 Err                               S.Err                |                                             
|               Sum/Avg               |                333                                33341                |               98.0                                 1.0                                1.1                                 0.5                                 2.6                                55.0                |                                             
write a WER result in exp/train_si284_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu4_bs16_lr10.0_warmup25000_mli512_mlo150_epochs100_accum1_lennormFalse_lsmunigram0.1/decode_test_eval92_beam10_emodel.last10.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result.wrd.txt                  
|               SPKR                  |                # Snt                              # Wrd                |                Corr                                 Sub                                  Del                                 Ins                                  Err                               S.Err                |                                         
|               Sum/Avg               |                 333                                5643                |                95.3                                 4.2                                  0.5                                 1.0                                  5.7                                46.2                | 
2019-04-22 18:47:54,630 (json2trn:44) INFO: reading data/lang_1char/train_si284_units.txt                                    
write a CER (or TER) result in exp/train_si284_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu4_bs16_lr10.0_warmup25000_mli512_mlo150_epochs100_accum1_lennormFalse_lsmunigram0.1/decode_test_dev93_beam10_emodel.last10.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result.txt
|               SPKR                  |               # Snt                               # Wrd               |               Corr                                 Sub                                 Del                                 Ins                                Err                               S.Err                |
|               Sum/Avg               |                503                                48634               |               96.9                                 1.3                                 1.8                                 0.8                                3.8                                65.0                |
write a WER result in exp/train_si284_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu4_bs16_lr10.0_warmup25000_mli512_mlo150_epochs100_accum1_lennormFalse_lsmunigram0.1/decode_test_dev93_beam10_emodel.last10.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_1layer_unit1000_sgd_bs300_word65000/result.wrd.txt
|               SPKR                  |                # Snt                              # Wrd                |               Corr                                  Sub                                 Del                                 Ins                                  Err                               S.Err                |
|               Sum/Avg               |                 503                                8234                |               92.6                                  6.5                                 0.9                                 1.4                                  8.8                                58.3                |
```

 However, I found running 100 epochs results much better in my previous impl.

**Comment by ShigekiKarita** [comment]
Let me rerun Transformer and chime recipe to check this merge for a while...

**Comment by ShigekiKarita** [comment]
https://github.com/espnet/espnet/pull/690/files#diff-740d366d6ac5d4770ff36bb06faa99cc

I updated pytorch RESULTS with 1 GPU and 4 GPU. Both settings are basically similar in terms of effective batchsize (64 utt.) except for ngpu but results are little bit different. 

**Training Tips:** While WSJ setting seems to be useful except for `lr_init`. I found `lr_init=1.0` and same other setting showed better performance in other recipes (e.g., CSJ, TEDLIUM, etc). Maybe larger dataset can allow larger minibatch for the best performance. Anyway I would not like to touch other recipes in this PR.

**Comment by ShigekiKarita** [comment]
I believe `egs/wsj/asr2` and `egs/chime4/asr_multich` work well after the merge commit.

**Comment by ShigekiKarita** [comment]
@sw005320 @kan-bayashi can you review and merge this?

**Comment by ShigekiKarita** [comment]
Sorry, I found some bugs in the last "enhance" stage of multi ch recipe. Please wait.

File: egs/wsj/asr2/run.sh [hunk 1]
```diff
@@ -318,8 +318,8 @@ fi
 
 if [ ${stage} -le 5 ] && [ ${stop_stage} -ge 5 ]; then
     echo "stage 5: Decoding"
-    nj=32
-    if [ ${n_average} -gt 1 ]; then
+    nj=${n_decode_job}
+    if [ ${n_average} -gt 1 ] && [ ! -e ${expdir}/results/model.last${n_average}.avg.best ]; then
```
Comment (Fhrozen) [inline egs/wsj/asr2/run.sh]: I am wondering about this part.
cause if the `avg.best` exists, then `recog_model` var will not change. 
If that is the case isn't better to put a second if inside this part and before `average_checkpoints.py`?

Comment (ShigekiKarita) [reply egs/wsj/asr2/run.sh]: Thanks. I think it is better to remove this line from here. I just added this `[ ! -e ${expdir}/results/model.last${n_average}.avg.best ]` when I test consistency of decoding on refactoring.

**Comment by ShigekiKarita** [comment]
I think the enhanced spectrogram plot issue is fixed

**Comment by sw005320** [comment]
> I think the enhanced spectrogram plot issue is fixed

Oh, thanks!

**Comment by ShigekiKarita** [comment]
@sw005320 can you merge this?

File: espnet/nets/asr_interface.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+class ASRInterface(object):
+    """ASR Interface for ESPnet model implementation
+
+    For chainer backend, __call__ method is needed.
```
Comment (sw005320) [inline espnet/nets/asr_interface.py]: Please modify this document.
Now chainer also uses the` forward` method, right?

**Comment by sw005320** [comment]
@kan-bayashi, please also review this PR.

**Comment by sw005320** [comment]
Please also update the main `README.md`. This is a notable addition.

File: espnet/nets/asr_interface.py (hunk 2 above)
Comment (ShigekiKarita) [reply espnet/nets/asr_interface.py]: Right.

**Comment by ShigekiKarita** [comment]
@sw005320 

>     * can we keep to use both pytorch and chainer run.sh by putting chainer run.sh to the local directory (e.g., `local/run_chainer.sh`)?

I agree. Chainer and pytorch still have difference (hyperparameters and joint decoding). I will revert the @Fhrozen recipe for chainer.

>     * I like `asr_interface.py` idea. Do you think ` espnet/nets` is an appropriate directory? Maybe ` espnet/asr`? We should also do the same thing for TTS (by @kan-bayashi?).

Hmm, I think the directory `espnet/nets` is OK because
- `espnet/net` is a root of ASR and TTS model definitions. It already contains some task specific implementations.
- `ASRInterface` (or `TTSInterface`) is a parent class of `espnet/net` models. Therefore, it is easy to find this class in the parent module.

>     * we may also refactor rnn based network files by moving them to `espnet/nets/pytorch_backend/rnn/`. @Fhrozen, please also follow @ShigekiKarita's style and make your directory structure consistent with pytorch_backend.
It is good idea. Maybe should we `e2e_asr.py` rename to `e2e_asr_rnn.py`?

**Comment by kan-bayashi** [comment]
Sorry for late review.
I will do it today.
And now I am testing it with CSJ recipe (This is just my curious).

File: espnet/asr/pytorch_backend/asr.py [hunk 3]
```diff
@@ -140,11 +142,13 @@ def update_core(self):
 
         # Compute the loss at this time step and accumulate it
         optimizer.zero_grad()  # Clear the parameter gradients
```
Comment (kan-bayashi) [inline espnet/asr/pytorch_backend/asr.py]: Is it OK to clear gradient here?

**Comment by ShigekiKarita** [comment]
@kan-bayashi Thanks for your comment. Let me keep some TODOs here
- [ ] justify the clear grad (add some comment or test)
- [ ] move `asr1` model under `espnet.nets.rnn` module
- [ ] checkout @Fhrozen 's chainer recipe and create `local/run_chainer.sh` and `local/run_pytorch.sh`

@sw005320 @Fhrozen @kan-bayashi Do you think `local/run_xxx.sh` should be like kaldi `local/chain/run_tdnn.sh`? You can see that the `run.sh` does **not** provide args to it because forwarding the cmd args is annoying.

https://github.com/kaldi-asr/kaldi/blob/35f96db7082559a57dcc222218db3f0be6dd7983/egs/wsj/s5/run.sh#L325

**Comment by ShigekiKarita** [comment]
@kan-bayashi  I also have tried CSJ without data augumentation (I only changed `lr_init=1.0`) but it actually results better than the current `asr1` recipe. Have you try data augumentation?

**Comment by kan-bayashi** [comment]
> it actually results better than the current asr1 recipe. 

That’s nice!

> Have you try data augumentation?

Now I’m training transformer with speed perturbation.
Once finished, I will report the results.

**Comment by kan-bayashi** [comment]
Sorry, maybe I do not understand.
Why do we need to separate `run_pytorch.sh` and `run_chainer.sh`?

**Comment by ShigekiKarita** [comment]
Unfortunately chainer does not result the consistent performance to pytorch
with the same setting at this time. Therefore, we decided to split them for
a while.

2019年5月15日(水) 7:28 Tomoki Hayashi <notifications@github.com>:

**Comment by kan-bayashi** [comment]
I see. 
I think that kaldi-style is ok for me.

File: espnet/asr/pytorch_backend/asr.py (hunk 3 above)
Comment (ShigekiKarita) [reply espnet/asr/pytorch_backend/asr.py]: wow I was wrong

Comment (kan-bayashi) [reply espnet/asr/pytorch_backend/asr.py]: Maybe this is the reason of different results when using different #ngpus #accums.

**Comment by sw005320** [comment]
I think #722 will fix the issue of maintaining two `run.sh` (we can just use two config files). For now, we can just keep your way.

**Comment by ShigekiKarita** [comment]
#722 sounds cool. I will ask @kamo-naoyuki  later. Anyway I'm still working on tensorboard support. It does not seem to be difficult. I will finish this PR today.

**Comment by ShigekiKarita** [comment]
Tensorboard support is done.  It works pretty good
image

image

I'm rerunning everything from scratch. Just a moment.

**Comment by ShigekiKarita** [comment]
I confirmed the RESULT of rerunning with `local/run_pytorch.sh` (1GPU, big model) is almost same to my first report.  I found `model.zero_grad()` bug was inserted during merging master. I'm still not sure if the `patience=3` is the best.  Anyway, this result is already faster (5 => 4 hours) and better (WER 7.1 => 4.8) than vggblstmp in `asr1`. 

I believe this PR is ready for master. @sw005320

**Comment by kan-bayashi** [comment]
I checked CSJ results without any tuning except for #epochs (early stopped at epoch 22 / 25).
It looks reasonable.

```
$ grep -e Avg -e SPKR exp/train_nodup_sp_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu2_bs32_lr10.0_warmup25000_mli512_mlo150_epochs25_accum2_lennormFalse_lsmunigram0.1/decode_eval*/result.txt
exp/train_nodup_sp_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu2_bs32_lr10.0_warmup25000_mli512_mlo150_epochs25_accum2_lennormFalse_lsmunigram0.1/decode_eval1_beam10_emodel.last5.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_2layer_unit650_sgd_bs256/result.txt:|              SPKR                   |              # Snt                             # Wrd               |              Corr                               Sub                              Del                               Ins                               Err                             S.Err               |
exp/train_nodup_sp_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu2_bs32_lr10.0_warmup25000_mli512_mlo150_epochs25_accum2_lennormFalse_lsmunigram0.1/decode_eval1_beam10_emodel.last5.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_2layer_unit650_sgd_bs256/result.txt:|              Sum/Avg                |              1272                              43897               |              94.7                               3.7                              1.6                               1.8                               7.2                              59.4               |
exp/train_nodup_sp_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu2_bs32_lr10.0_warmup25000_mli512_mlo150_epochs25_accum2_lennormFalse_lsmunigram0.1/decode_eval2_beam10_emodel.last5.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_2layer_unit650_sgd_bs256/result.txt:|              SPKR                   |              # Snt                             # Wrd               |              Corr                               Sub                              Del                               Ins                               Err                             S.Err               |
exp/train_nodup_sp_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu2_bs32_lr10.0_warmup25000_mli512_mlo150_epochs25_accum2_lennormFalse_lsmunigram0.1/decode_eval2_beam10_emodel.last5.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_2layer_unit650_sgd_bs256/result.txt:|              Sum/Avg                |              1292                              43623               |              95.5                               3.3                              1.3                               1.2                               5.7                              60.1               |
exp/train_nodup_sp_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu2_bs32_lr10.0_warmup25000_mli512_mlo150_epochs25_accum2_lennormFalse_lsmunigram0.1/decode_eval3_beam10_emodel.last5.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_2layer_unit650_sgd_bs256/result.txt:|              SPKR                   |              # Snt                             # Wrd               |              Corr                               Sub                              Del                               Ins                               Err                             S.Err               |
exp/train_nodup_sp_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu2_bs32_lr10.0_warmup25000_mli512_mlo150_epochs25_accum2_lennormFalse_lsmunigram0.1/decode_eval3_beam10_emodel.last5.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0_2layer_unit650_sgd_bs256/result.txt:|              Sum/Avg                |              1385                              28225               |              95.3                               3.3                              1.4                               1.9                               6.6                              42.0               |

```

**Comment by ShigekiKarita** [comment]
Cool. I ran the same setting except  `--patience 0 --epochs 100 --n-average 10` even if it looks like converged in epoch 22. It would still improve its acc around epoch 95. This is my result

- eval1 CER 6.5
- eval2 CER 4.5
- eval3 CER 5.2

**Comment by ShigekiKarita** [comment]
In fact, I also tried to tune decoding params with train_dev set (our valid set). It did not improve CER/WER.

**Comment by kan-bayashi** [comment]
I checked the performance with current `csj/asr1` decoding parameters.
That is quite nice :)

```
write a CER (or TER) result in exp/train_nodup_sp_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu2_bs32_lr10.0_warmup25000_mli512_mlo150_epochs25_accum2_lennormFalse_lsmunigram0.1/decode_eval2_beam10_emodel.last5.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm0.3_2layer_unit650_sgd_bs256/result.txt
|              SPKR                   |              # Snt                             # Wrd               |              Corr                               Sub                              Del                               Ins                               Err                             S.Err               |
|              Sum/Avg                |              1292                              43623               |              96.1                               2.6                              1.3                               0.6                               4.5                              53.9               |
--
write a CER (or TER) result in exp/train_nodup_sp_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu2_bs32_lr10.0_warmup25000_mli512_mlo150_epochs25_accum2_lennormFalse_lsmunigram0.1/decode_eval1_beam10_emodel.last5.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm0.3_2layer_unit650_sgd_bs256/result.txt
|              SPKR                   |              # Snt                             # Wrd               |              Corr                               Sub                              Del                               Ins                               Err                             S.Err               |
|              Sum/Avg                |              1272                              43897               |              95.0                               3.3                              1.8                               0.8                               5.9                              55.4               |
--
write a CER (or TER) result in exp/train_nodup_sp_pytorch_transformer_conv2d_e12_unit2048_d6_unit2048_aheads4_dim256_mtlalpha0.3_noam_sampprob0.0_ngpu2_bs32_lr10.0_warmup25000_mli512_mlo150_epochs25_accum2_lennormFalse_lsmunigram0.1/decode_eval3_beam10_emodel.last5.avg.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm0.3_2layer_unit650_sgd_bs256/result.txt
|              SPKR                   |              # Snt                             # Wrd               |              Corr                               Sub                              Del                               Ins                               Err                             S.Err               |
|              Sum/Avg                |              1385                              28225               |              95.9                               2.8                              1.3                               0.9                               5.0                              38.7               |
```

**Comment by kan-bayashi** [comment]
After merge this PR, I will make PR of `csj/asr2` with the RESULTS.

**Comment by kan-bayashi** [comment]
> even if it looks like converged in epoch 22. It would still improve its acc around epoch 95. 

I used speed perturbation to augment training data.
I will check the effect of further training.

**Comment by ShigekiKarita** [comment]
new decoding result looks awesome!! cannot wait the result with data aug!!

**Comment by ShigekiKarita** [comment]
Oh, my result seems to use the csj/asr1 decoding setting. Therefore it can be considered as following

| system | eval1 | eval2 | eval3 |
|:---|---:|---:|---:|
|1GPU baseline @ShigekiKarita  | 6.5 | 4.5 | 5.2 |
|+speed perturb @kan-bayashi  | 5.9 | 4.5 | 5.0 |
|+spectaug @kan-bayashi | ? | ? | ? |

**Pull request merged**